etcdserver/api/v3rpc/interceptor.go

```diff
@@ -231,8 +232,13 @@ func newStreamInterceptor(s *etcdserver.EtcdServer) grpc.StreamServerInterceptor
 					return rpctypes.ErrGRPCNoLeader
 				}
 
-				cctx, cancel := context.WithCancel(ss.Context())
-				ss = serverStreamWithCtx{ctx: cctx, cancel: &cancel, ServerStream: ss}
+				cancelCtx, cancelFn := context.WithCancel(ss.Context())
+				monitorCtx := &leaderMonitoringContext{
```

Comment:
Do we really need this struct to wrap around the error? I am seeing the cancel func is only called in `monitorLeader` `(*ssWithCtx.cancel)()`. Maybe simpler to embed this error info in `serverStreamWithCtx` and implement `Err` method against it?

Reply:
I'm not sure I understand your proposal. The `watchServer.Watch` method (in `v3rpc/watch.go`) handles the context cancellation and has only the context itself and its embedded error to interpret (via `Context.Err()`). The stdlib doesn't seem to provide any context implementations that allow errors to be injected, hence my wrapper. 

Would you be able to try explaining in more detail what you had in mind?

Reply:
@gyuho friendly ping — I think what I have here solves the problem, and I'm not entirely clear on what the proposed alternative is. Can you please clarify?

Reply:
@ironcladlou Sorry for delay. I will take another look today.

Reply:
> embedded error to interpret (via Context.Err()). The stdlib doesn't seem to provide any context implementations that allow errors to be injected, hence my wrapper.

Would combining these to `serverStreamWithCtx` make this simpler?

```
func (ssc serverStreamWithCtx) Err() error {
	c.lock.RLock()
	defer c.lock.RUnlock()
	if c.cancelReason != nil {
		return c.cancelReason
	}
	return c.Context.Err()
}
```

Reply:
@gyuho okay, I think I now see where you were going with this suggestion. I spent some more time thinking about it and came up with a slightly new factoring which I hope is easier to understand.

I greatly simplified the context responsibility with a `cancellableContext` struct which I hope now has very clear and self contained semantics. Then I made `serverStreamWithCtx` depend on a concrete `cancellableContext` and was able to move the specific cancellation error to the point it's actually generated instead of indirectly way out in the setup code like before.

I can still imagine a factoring where `cancellableContext` is moved into `serverStreamWithCtx` which itself would implement `context.Context`, but I currently think with the new factoring the `cancellableContext` is clear and self contained enough to keep it separate.

Let me know what you think — I really appreciate the feedback!

Reply:
It looks clear to me.
I would have some doubts if this 'alternative' context was a public class, but 
as long it stays private implementation detail (as in this case), I think its good.